README.rst
.. image:: https://travis-ci.com/zopefoundation/zLOG.svg?branch=master
        :target: https://travis-ci.com/zopefoundation/zLOG

.. image:: https://coveralls.io/repos/github/zopefoundation/zLOG/badge.svg?branch=master
        :target: https://coveralls.io/github/zopefoundation/zLOG?branch=master

.. image:: https://img.shields.io/pypi/v/zLOG.svg
        :target: https://pypi.org/project/zLOG/
        :alt: Current version on PyPI

.. image:: https://img.shields.io/pypi/pyversions/zLOG.svg
        :target: https://pypi.org/project/zLOG/
        :alt: Supported Python versions

``zLOG`` 
========

This package provides a general logging facility that, at this point,
is just a small shim over Python's logging module.  Therefore, unless
you need to support a legacy package from the Zope 2 world, you're
probably better off using Python's logging module.


CHANGES.rst
Changelog
=========

4.0 (unreleased)
----------------

- Add support for Python 3.10, 3.11.

- Drop support for Python 2.7, 3.5, 3.6.


3.1 (2020-11-18)
----------------

- Add support for Python 3.6 up to 3.9.

- Drop support for Python 3.3 and 3.4.

- Flake8 the code.


3.0 (2016-04-03)
----------------

- Add coverage testing.

- Add Python 3.3 - 3.5 compatibility.

- Declare currently-supported Python versions, and test them.

- Normalize package structure (``README.rst``, ``CHANGES.rst``). Synthesize
  package description from README.rst and CHANGES.rst.

- Use nose for testing instead of zope.testrunner and test
  for 100% test coverage.

2.12.0 (2012-08-30)
-------------------

- Rely on refactored base class from ``ZConfig`` for testing logging.


src/zLOG/tests/testzLog.py
##############################################################################
#
# Copyright (c) 2001, 2002 Zope Foundation and Contributors.
# All Rights Reserved.
#
# This software is subject to the provisions of the Zope Public License,
# Version 2.1 (ZPL).  A copy of the ZPL should accompany this distribution.
# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
# WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
# FOR A PARTICULAR PURPOSE.
#
##############################################################################

import logging
import os
import sys
import tempfile
import time
import unittest

import zLOG


severity_string = {
    -300: 'TRACE',
    -200: 'DEBUG',
    -100: 'BLATHER',
    0: 'INFO',
    100: 'PROBLEM',
    200: 'ERROR',
    300: 'PANIC',
}


class EventLogTest(unittest.TestCase):
    """Test zLOG with the default implementation."""

    def setUp(self):
        self.f, self.path = tempfile.mkstemp()
        self._severity = 0
        self.loghandler = self.setLog()

    def tearDown(self):
        self.loghandler.close()
        os.close(self.f)
        os.remove(self.path)

    def setLog(self, severity=0):
        logger = logging.getLogger('basic')
        logger.setLevel(logging.DEBUG)
        formatter = logging.Formatter(
            fmt='------\n%(asctime)s %(name)s %(levelname)s %(message)s',
            datefmt='%Y-%m-%dT%H:%M:%S')
        handler = logging.FileHandler(self.path)
        handler.setFormatter(formatter)
        logger.addHandler(handler)

        self._severity = severity
        return handler

    def verifyEntry(self, f, time=None, subsys=None, severity=None,
                    summary=None, detail=None, error=None):
        # skip to the beginning of next entry
        line = f.readline().strip()
        line = f.readline().strip()
        _time, rest = line.split(" ", 1)
        if subsys is not None:  # pragma: no cover
            self.assertIn(subsys, rest, "subsystem mismatch")
        if severity is not None and severity >= self._severity:
            s = severity_string[severity]
            self.assertIn(s, rest, "severity mismatch")
        if summary is not None:
            self.assertIn(summary, rest, "summary mismatch")
        if detail is not None:
            line = f.readline()
            self.assertNotEqual(line.find(detail), -1, "missing detail")
        if error is not None:
            line = f.readline().strip()
            self.assertTrue(line.startswith('Traceback'),
                            "missing traceback")

    def getLogFile(self):
        return open(self.path)

    def test_basics(self):
        zLOG.LOG("basic", zLOG.INFO, "summary")
        with self.getLogFile() as f:
            self.verifyEntry(f, subsys="basic", summary="summary")

    def test_detail(self):
        zLOG.LOG("basic", zLOG.INFO, "xxx", "this is a detail")
        with self.getLogFile() as f:
            self.verifyEntry(f, subsys="basic", detail="detail")

    def test_error(self):
        try:
            1 / 0
        except ZeroDivisionError:
            err = sys.exc_info()

        zLOG.LOG("basic", zLOG.INFO, "summary")
        zLOG.LOG("basic", zLOG.ERROR, "raised exception", error=err)
        with self.getLogFile() as f:
            self.verifyEntry(f, subsys="basic", summary="summary")
            self.verifyEntry(f, subsys="basic", severity=zLOG.ERROR, error=err)

    def test_reraise_error(self):
        self.setLog()
        try:
            1 / 0
        except ZeroDivisionError:
            err = sys.exc_info()

        self.assertRaises(ZeroDivisionError, zLOG.LOG, "basic", zLOG.ERROR,
                          "raised exception", error=err, reraise=True)
        with self.getLogFile() as f:
            self.verifyEntry(f, subsys="basic", severity=zLOG.ERROR, error=err)

    def test_bbb(self):
        """Test existence of BBB methods that do nothing."""
        zLOG.initialize()
        zLOG.set_initializer(lambda: False)  # pragma: no cover
        zLOG.register_subsystem('foo')
        self.assertTrue('foo' in zLOG._subsystems)

    def test_severity_string(self):
        # severity in mapping
        self.assertEqual(zLOG.severity_string(100), 'PROBLEM(100)')
        # severity not in mapping
        self.assertEqual(zLOG.severity_string(99), '(99)')

    def test_log_time(self):
        self.assertTrue(zLOG.log_time().startswith(
            '%4.4d-%2.2d-%2.2dT' % time.localtime()[:3]))


src/zLOG/tests/test_logging.py
##############################################################################
#
# Copyright (c) 2004 Zope Foundation and Contributors.
# All Rights Reserved.
#
# This software is subject to the provisions of the Zope Public License,
# Version 2.1 (ZPL).  A copy of the ZPL should accompany this distribution.
# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
# WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
# FOR A PARTICULAR PURPOSE.
#
##############################################################################
"""Tests of the integration with the standard logging package."""

import logging
import unittest

from ZConfig.components.logger.tests.support import LoggingTestHelper

import zLOG
from zLOG.EventLogger import log_write


class CollectingHandler(logging.Handler):

    def __init__(self):
        logging.Handler.__init__(self)
        self.records = []

    def emit(self, record):
        self.records.append(record)


class LoggingIntegrationTestCase(LoggingTestHelper, unittest.TestCase):

    def setUp(self):
        LoggingTestHelper.setUp(self)
        self.handler = CollectingHandler()
        self.records = self.handler.records
        self.logger = logging.getLogger()
        self.logger.addHandler(self.handler)

    def test_log_record(self):
        # log_write(subsystem, severity, summary, detail, error)
        log_write("sample.subsystem", zLOG.WARNING, "summary", "detail", None)
        self.assertEqual(len(self.records), 1)
        record = self.records[0]
        self.assertEqual(record.levelno, logging.WARN)
        self.assertEqual(record.name, "sample.subsystem")
        # Make sure both the message and the detail information appear
        # in the text that gets logged:
        record.msg.index("summary")
        record.msg.index("detail")


src/zLOG/EventLogger.py
##############################################################################
#
# Copyright (c) 2001, 2002 Zope Foundation and Contributors.
# All Rights Reserved.
#
# This software is subject to the provisions of the Zope Public License,
# Version 2.1 (ZPL).  A copy of the ZPL should accompany this distribution.
# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
# WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
# FOR A PARTICULAR PURPOSE
#
##############################################################################
"""
A logging module which handles event messages.

This uses Vinay Sajip's PEP 282 logging module.
"""

import logging
import time


# XXX Defining custom levels needs to move to
# ZConfig.components.logger, I think.  It only needs to be done when
# logging configuration is activated (the logger factory is called).
#
# Custom logging levels
CUSTOM_BLATHER = 15  # Mapping for zLOG.BLATHER
CUSTOM_TRACE = 5  # Mapping for zLOG.TRACE
logging.addLevelName("BLATHER", CUSTOM_BLATHER)
logging.addLevelName("TRACE", CUSTOM_TRACE)


def fmt_raise(error):
    raise error[0](error[1]).with_traceback(error[2])


def log_write(subsystem, severity, summary, detail, error):
    level = (zlog_to_pep282_severity_cache_get(severity) or
             zlog_to_pep282_severity(severity))

    msg = summary
    if detail:
        msg = "{}\n{}".format(msg, detail)

    logger = logging.getLogger(subsystem)
    logger.log(level, msg, exc_info=error)


def severity_string(severity, mapping={
        -300: 'TRACE',
        -200: 'DEBUG',
        -100: 'BLATHER',
        0: 'INFO',
        100: 'PROBLEM',
        200: 'ERROR',
        300: 'PANIC'}):
    """Convert a severity code to a string."""
    s = mapping.get(int(severity), '')
    return "{}({})".format(s, severity)


def zlog_to_pep282_severity(zlog_severity):
    """We map zLOG severities to PEP282 severities here.

    This is how they are mapped:

    zLOG severity                      PEP282 severity
    -------------                      ---------------
    PANIC (300)                        FATAL, CRITICAL (50)
    ERROR (200)                        ERROR (40)
    WARNING, PROBLEM (100)             WARN (30)
    INFO (0)                           INFO (20)
    BLATHER (-100)                     BLATHER (15) [*]
    DEBUG (-200)                       DEBUG (10)
    TRACE (-300)                       TRACE (5) [*]

    [*] BLATHER and TRACE are custom logging levels.
    """
    sev = zlog_severity
    if sev >= 300:
        return logging.FATAL
    if sev >= 200:
        return logging.ERROR
    if sev >= 100:
        return logging.WARN
    if sev >= 0:
        return logging.INFO
    if sev >= -100:
        return CUSTOM_BLATHER
    if sev >= -200:
        return logging.DEBUG
    return CUSTOM_TRACE


zlog_to_pep282_severity_cache = {}
for _sev in range(-300, 301, 100):
    zlog_to_pep282_severity_cache[_sev] = zlog_to_pep282_severity(_sev)
zlog_to_pep282_severity_cache_get = zlog_to_pep282_severity_cache.get


def log_time():
    """Return a simple time string without spaces suitable for logging."""
    return ("%4.4d-%2.2d-%2.2dT%2.2d:%2.2d:%2.2d"
            % time.localtime()[:6])


src/zLOG/__init__.py
##############################################################################
#
# Copyright (c) 2001, 2002 Zope Foundation and Contributors.
# All Rights Reserved.
#
# This software is subject to the provisions of the Zope Public License,
# Version 2.1 (ZPL).  A copy of the ZPL should accompany this distribution.
# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
# WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
# FOR A PARTICULAR PURPOSE.
#
##############################################################################

"""General logging facility

Note:
  This module exists only for backward compatibility.  Any new code
  for Zope 2.8 and newer should use the logging module from Python's
  standard library directly.  zLOG is only an API shim to map existing
  use of zLOG onto the standard logging API.

This module attempts to provide a simple programming API for logging
with a pluggable API for defining where log messages should go.

Programmers call the LOG function to log information.

The LOG function, in turn, calls the log_write method to actually write
logging information somewhere.  This module provides a very simple
log_write implementation.  It is intended that applications main
programs will replace this method with a method more suited to their needs.

The module provides a register_subsystem method that does nothing, but
provides a hook that logging management systems could use to collect
information about subsystems being used.

The module defines several standard severities:

  TRACE=-300   -- Trace messages

  DEBUG=-200   -- Debugging messages

  BLATHER=-100 -- Somebody shut this app up.

  INFO=0       -- For things like startup and shutdown.

  PROBLEM=100  -- This isn't causing any immediate problems, but deserves
                  attention.

  WARNING=100  -- A wishy-washy alias for PROBLEM.

  ERROR=200    -- This is going to have adverse effects.

  PANIC=300    -- We're dead!

To plug in a log handler, simply replace the log_write function
with a callable object that takes 5 arguments:

      subsystem -- The subsystem generating the message (e.g. ZODB)

      severity -- The "severity" of the event.  This may be an integer or
                  a floating point number.  Logging back ends may
                  consider the int() of this valua to be significant.
                  For example, a backend may consider any severity
                  whos integer value is WARNING to be a warning.

      summary -- A short summary of the event

      detail -- A detailed description

      error -- A three-element tuple consisting of an error type, value, and
               traceback.  If provided, then a summary of the error
               is added to the detail.

The default logging facility uses Python's logging module as a
back-end; configuration of the logging module must be handled
somewhere else.

"""

from traceback import format_exception  # noqa: F401

from zLOG.EventLogger import fmt_raise
from zLOG.EventLogger import log_time
from zLOG.EventLogger import log_write
from zLOG.EventLogger import severity_string  # noqa: F401


# Standard severities
TRACE = -300
DEBUG = -200
BLATHER = -100
INFO = 0
PROBLEM = 100
WARNING = 100
ERROR = 200
PANIC = 300


def initialize():
    pass


def set_initializer(func):
    """Set the function used to re-initialize the logs.

    This should be called when initialize_from_environment() is not
    appropiate.

    This does not ensure that the new function gets called; the caller
    should do that separately.
    """
    pass


def LOG(subsystem, severity, summary, detail='', error=None, reraise=None):
    """Log some information

    The required arguments are:

      subsystem -- The subsystem generating the message (e.g. ZODB)

      severity -- The "severity" of the event.  This may be an integer or
                  a floating point number.  Logging back ends may
                  consider the int() of this value to be significant.
                  For example, a backend may consider any severity
                  whos integer value is WARNING to be a warning.

      summary -- A short summary of the event

      detail -- A detailed description

      error -- A three-element tuple consisting of an error type, value, and
               traceback.  If provided, then a summary of the error
               is added to the detail.

      reraise -- If provided with a true value, then the error given by
                 error is reraised.

    """
    log_write(subsystem, severity, summary, detail, error)
    if reraise and error:
        raise fmt_raise(error)


_subsystems = []


def register_subsystem(subsystem):
    """Register a subsystem name

    A logging facility might replace this function to collect information about
    subsystems used in an application.
    """
    _subsystems.append(subsystem)


# Most apps interested in logging only want the names below.
__all__ = ['LOG', 'TRACE', 'DEBUG', 'BLATHER', 'INFO', 'PROBLEM',
           'WARNING', 'ERROR', 'PANIC', 'log_time']
